File: backend/src/modules/ai-mentor/rule-based-extractor.ts
/**
 * Rule-Based Module Extractor — Fallback when Ollama is offline.
 *
 * Splits raw PDF text into structured lesson content blocks using
 * deterministic heuristics (regex patterns, paragraph boundaries).
 *
 * Produces the same JSON shape as the Ollama-based extraction so
 * downstream code (review UI, apply endpoint) works identically.
 *
 * This is NOT an @Injectable — it's a pure utility function.
 */

// ────────────────────────────────────────────────────────────────────────────
// Types (matches the Ollama output contract)
// ────────────────────────────────────────────────────────────────────────────

export interface ExtractedBlock {
  type: 'text' | 'question' | 'divider';
  order: number;
  content: Record<string, unknown>;
  metadata: Record<string, unknown>;
}

export interface ExtractedLesson {
  title: string;
  description: string;
  blocks: ExtractedBlock[];
}

export interface ExtractionResult {
  title: string;
  description: string;
  lessons: ExtractedLesson[];
}

// ────────────────────────────────────────────────────────────────────────────
// Heading / section detection patterns
// ────────────────────────────────────────────────────────────────────────────

/**
 * Matches common DepEd-module / textbook section markers:
 *   "Lesson 1: Introduction"
 *   "CHAPTER 3 — Fractions"
 *   "Module 2: Data Handling"
 *   "I. Objectives"
 *   "A. Motivation"
 */
const HEADING_RE =
  /^(?:(?:lesson|chapter|module|unit|topic|part)\s*\d*\s*[:\-–—.]\s*(.+)|([IVXLCDM]+\.\s+.+)|([A-Z]\.\s+.+))/im;

/** Matches quiz / test question patterns like  "1. What is…?"  or  "1) …" */
const QUESTION_LINE_RE = /^\s*\d+[.)]\s+.+\??\s*$/;

/** Matches multiple-choice options like  "A) …"  "a. …"  "(a) …" */
const OPTION_RE = /^\s*[\(\[]?[a-dA-D][.):\]]\s*.+/;

// ────────────────────────────────────────────────────────────────────────────
// Main extraction function
// ────────────────────────────────────────────────────────────────────────────

export function extractWithRules(rawText: string): ExtractionResult {
  // ── 1. Normalise whitespace ────────────────────────────────────────────
  const cleaned = rawText
    .replace(/\r\n/g, '\n')        // Windows → Unix line endings
    .replace(/\f/g, '\n\n')        // form-feed (PDF page breaks) → double newline
    .replace(/[ \t]+/g, ' ')       // collapse horizontal whitespace
    .replace(/\n{3,}/g, '\n\n');   // max two consecutive newlines

  // ── 2. Split into sections by heading patterns ─────────────────────────
  const lines = cleaned.split('\n');

  interface RawSection {
    heading: string;
    bodyLines: string[];
  }

  const sections: RawSection[] = [];
  let current: RawSection = { heading: 'Untitled Section', bodyLines: [] };

  for (const line of lines) {
    const match = HEADING_RE.exec(line.trim());
    if (match) {
      // Push previous section if it has content
      if (current.bodyLines.length > 0) {
        sections.push(current);
      }
      current = {
        heading: (match[1] || match[2] || match[3] || line).trim(),
        bodyLines: [],
      };
    } else {
      current.bodyLines.push(line);
    }
  }
  // Push the last section
  if (current.bodyLines.length > 0) {
    sections.push(current);
  }

  // If no headings were detected, treat the entire text as one section
  if (sections.length === 0) {
    sections.push({ heading: 'Extracted Content', bodyLines: lines });
  }

  // ── 3. Convert sections → lessons with content blocks ──────────────────
  const lessons: ExtractedLesson[] = sections.map((section) => {
    const blocks = buildBlocks(section.bodyLines);
    return {
      title: section.heading,
      description: '',
      blocks,
    };
  });

  // ── 4. Derive a top-level title from the first non-empty line ──────────
  const firstNonEmpty = lines.find((l) => l.trim().length > 0) || 'Extracted Module';

  return {
    title: firstNonEmpty.trim().substring(0, 200),
    description: `Auto-extracted (rule-based) from PDF — ${sections.length} section(s) detected.`,
    lessons,
  };
}

// ────────────────────────────────────────────────────────────────────────────
// Helpers
// ────────────────────────────────────────────────────────────────────────────

/**
 * Converts an array of body lines into ordered content blocks.
 *
 * Strategy:
 *  1. Accumulate consecutive text lines into a single `text` block.
 *  2. When a question pattern is detected, flush the text block and start
 *     accumulating question + option lines into a `question` block.
 *  3. Insert a `divider` between large content gaps (double blank lines).
 */
function buildBlocks(bodyLines: string[]): ExtractedBlock[] {
  const blocks: ExtractedBlock[] = [];
  let order = 0;
  let textBuffer: string[] = [];
  let questionBuffer: string[] = [];
  let inQuestion = false;

  const flushText = () => {
    const joined = textBuffer.join('\n').trim();
    if (joined.length > 0) {
      blocks.push({
        type: 'text',
        order: order++,
        content: { text: joined },
        metadata: { source: 'rule-based' },
      });
    }
    textBuffer = [];
  };

  const flushQuestion = () => {
    const joined = questionBuffer.join('\n').trim();
    if (joined.length > 0) {
      blocks.push({
        type: 'question',
        order: order++,
        content: { text: joined },
        metadata: { source: 'rule-based', detectedAs: 'question-pattern' },
      });
    }
    questionBuffer = [];
    inQuestion = false;
  };

  for (const line of bodyLines) {
    const trimmed = line.trim();

    // Blank line → potential divider / section break
    if (trimmed.length === 0) {
      if (inQuestion) {
        flushQuestion();
      }
      continue; // We don't insert a divider for every blank line — only between headings
    }

    // Detect question line
    if (QUESTION_LINE_RE.test(trimmed)) {
      if (!inQuestion) {
        flushText(); // Flush any accumulated text before the question
      }
      inQuestion = true;
      questionBuffer.push(trimmed);
      continue;
    }

    // Detect option line (belongs to current question)
    if (inQuestion && OPTION_RE.test(trimmed)) {
      questionBuffer.push(trimmed);
      continue;
    }

    // If we were in a question but this line is regular text, flush the question
    if (inQuestion) {
      flushQuestion();
    }

    // Regular text line
    textBuffer.push(trimmed);
  }

  // Flush remaining buffers
  if (inQuestion) {
    flushQuestion();
  }
  flushText();

  return blocks;
}

File: ai-service/app/rule_based_extractor.py
"""
Rule-Based Module Extractor – Fallback when Ollama is offline.
"""

from __future__ import annotations

import re
from typing import Any

HEADING_RE = re.compile(
    r"^(?:"
    r"(?:lesson|chapter|module|unit|topic|part)\s*\d*\s*[:\-\u2013\u2014.]\s*(.+)"
    r"|([IVXLCDM]+\.\s+.+)"
    r"|([A-Z]\.\s+.+)"
    r")",
    re.I | re.M,
)

QUESTION_LINE_RE = re.compile(r"^\s*\d+[.)]\s+.+\??\s*$")
OPTION_RE = re.compile(r"^\s*[\(\[]?[a-dA-D][.):\]]\s*.+")


def extract_with_rules(
    raw_text: str,
    *,
    source_method: str = "rule_based",
    chunk_index: int | None = None,
    page_start: int | None = None,
    page_end: int | None = None,
    confidence: float = 0.45,
) -> dict[str, Any]:
    cleaned = raw_text.replace("\r\n", "\n").replace("\f", "\n\n")
    cleaned = re.sub(r"[ \t]+", " ", cleaned)
    cleaned = re.sub(r"\n{3,}", "\n\n", cleaned)

    lines = cleaned.split("\n")

    sections: list[dict[str, Any]] = []
    current: dict[str, Any] = {"heading": "Untitled Section", "body_lines": []}

    for line in lines:
        match = HEADING_RE.match(line.strip())
        if match:
            if current["body_lines"]:
                sections.append(current)
            current = {
                "heading": (match.group(1) or match.group(2) or match.group(3) or line).strip(),
                "body_lines": [],
            }
        else:
            current["body_lines"].append(line)

    if current["body_lines"]:
        sections.append(current)

    if not sections:
        sections.append({"heading": "Extracted Content", "body_lines": lines})

    lessons = []
    for section in sections:
        blocks = build_blocks_from_text(
            "\n".join(section["body_lines"]),
            source_method=source_method,
            chunk_index=chunk_index,
            page_start=page_start,
            page_end=page_end,
            confidence=confidence,
        )
        lessons.append({"title": section["heading"], "description": "", "blocks": blocks})

    first_non_empty = next((l.strip() for l in lines if l.strip()), "Extracted Module")

    return {
        "title": first_non_empty[:200],
        "description": f"Auto-extracted (rule-based) from PDF — {len(sections)} section(s) detected.",
        "lessons": lessons,
    }


def build_blocks_from_text(
    body_text: str,
    *,
    source_method: str = "rule-based",
    chunk_index: int | None = None,
    page_start: int | None = None,
    page_end: int | None = None,
    confidence: float = 0.6,
) -> list[dict[str, Any]]:
    return _build_blocks(
        body_text.split("\n"),
        source_method=source_method,
        chunk_index=chunk_index,
        page_start=page_start,
        page_end=page_end,
        confidence=confidence,
    )


def _build_blocks(
    body_lines: list[str],
    *,
    source_method: str = "rule-based",
    chunk_index: int | None = None,
    page_start: int | None = None,
    page_end: int | None = None,
    confidence: float = 0.6,
) -> list[dict[str, Any]]:
    blocks: list[dict[str, Any]] = []
    order = 0
    text_buffer: list[str] = []
    question_buffer: list[str] = []
    in_question = False

    provenance = {
        "source": source_method,
        "chunkIndex": chunk_index,
        "pageStart": page_start,
        "pageEnd": page_end,
        "confidence": round(confidence, 4),
    }

    def flush_text():
        nonlocal order
        joined = "\n".join(text_buffer).strip()
        if joined:
            blocks.append({
                "type": "text",
                "order": order,
                "content": {"text": joined},
                "metadata": dict(provenance),
            })
            order += 1
        text_buffer.clear()

    def flush_question():
        nonlocal order, in_question
        joined = "\n".join(question_buffer).strip()
        if joined:
            blocks.append({
                "type": "question",
                "order": order,
                "content": {"text": joined},
                "metadata": {**provenance, "detectedAs": "question-pattern"},
            })
            order += 1
        question_buffer.clear()
        in_question = False

    for line in body_lines:
        trimmed = line.strip()

        if not trimmed:
            if in_question:
                flush_question()
            continue

        if QUESTION_LINE_RE.match(trimmed):
            if not in_question:
                flush_text()
            in_question = True
            question_buffer.append(trimmed)
            continue

        if in_question and OPTION_RE.match(trimmed):
            question_buffer.append(trimmed)
            continue

        if in_question:
            flush_question()

        text_buffer.append(trimmed)

    if in_question:
        flush_question()
    flush_text()

    return blocks
